Lib/concurrent/futures/_base.py

```diff
@@ -170,6 +170,17 @@ def _create_and_install_waiters(fs, return_when):
 
     return waiter
 
+
+def _yield_future(fs, waiter, ref_collect=()):
+    while fs:
+        with fs[0]._condition:
+            fs[0]._waiters.remove(waiter)
+
+        for future_list in ref_collect:
+            future_list.remove(fs[0])
```

Comment:
In this use case future_list will be always a set. 
`set().remove` is O(1).
I will rename `future_list` to `futures_collection`